src/tree.ml

```diff
@@ -0,0 +1,724 @@
+    may be more than two occurrences of a variables. In that case we will need
+    to check convertibility between more than two terms.
+
+    Free variable constraints are used to verify which variables are free in a
+    term. If there is a rule of the form [f (λ x y, &Y[y]) → &Y], then we need
+    to check that the term at position [{0.0}] depends only on [y]. *)
+
+(** Module providing a representation for pools of conditions. *)
+module CP = struct
+  (** Functional sets of pairs of integers. *)
+  module PSet = Set.Make(
+    struct
+      type t = int * int
+
+      let compare : t -> t -> int = fun (i1,i2) (j1,j2) ->
+        match i1 - j1 with 0 -> i2 - j2 | k -> k
+    end)
+
+  (** A pool of (convertibility and free variable) conditions. *)
+  type t =
+    { nl_partial : int IntMap.t
+    (** An association [(e, v)] is a slot [e] of the [env] array with a slot
+          [v] of the [vars] array. *)
+    ; nl_conds : PSet.t
+    (** Set of convertibility constraints that can be checked. *)
+    ; fv_conds : (tvar array) IntMap.t
+    (** Set of free variable constraints. *) }
+
+  (** {b NOTE} the integer indices stored in {!field:nl_conds} and the keys of
+      in {!field:fv_conds} correspond to (valid) positions in the [vars] array
+      of the {!val:Eval.tree_walk} function. *)
+
+  (** [empty] is the condition pool holding no condition. *)
+  let empty : t =
+    { nl_partial = IntMap.empty
+    ; nl_conds = PSet.empty
+    ; fv_conds = IntMap.empty }
+
+  (** [pp oc pool] prints condition pool [pool] to channel [oc]. *)
+  let pp : t pp = fun oc pool ->
+    let pp_fv oc fv_conds =
+      let pp_sep oc () = Format.pp_print_string oc ";" in
+      let pp_tvs = Format.pp_print_list ~pp_sep Print.pp_tvar in
+      let ppit oc (a, b) =
+        Format.fprintf oc "@[(%a@@%d)@]" pp_tvs (Array.to_list b) a
+      in
+      let pp_sep oc () = Format.pp_print_string oc "::" in
+      Format.fprintf oc "@[<v>@[%a@]@]"
+        (Format.pp_print_list ~pp_sep ppit) (IntMap.bindings fv_conds)
+    in
+    let pp_partials oc partials =
+      let pp_sep oc _ = Format.pp_print_string oc ";" in
+      let pp_part oc (i, j) = Format.fprintf oc "@[(%d↦%d)@]" i j in
+      Format.fprintf oc "@[<h>%a@]" (Format.pp_print_list ~pp_sep pp_part)
+        (IntMap.bindings partials)
+    in
+    let pp_available oc available =
+      let pp_av oc (i, j) = Format.fprintf oc "@[%d≡%d@]" i j in
+      let pp_sep = Format.pp_print_cut in
+      Format.fprintf oc "@[<hov>%a@]" (Format.pp_print_list ~pp_sep pp_av)
+        (PSet.elements available)
+    in
+    Format.fprintf oc "@[<hov>%a/@,%a/@,%a@]" pp_fv pool.fv_conds
+      pp_partials pool.nl_partial pp_available pool.nl_conds
+
+  (** [is_empty pool] tells whether the pool of constraints is empty. *)
+  let is_empty : t -> bool = fun pool ->
+    PSet.is_empty pool.nl_conds && IntMap.is_empty pool.fv_conds
+
+  (* TODO cleanup from here. *)
```

Comment:
I'm gonna come back later for further cleaning up of the doc, but if you want to make a pass that would be great. For instance, the comment of `instantiate_nl` is not very helpful, and the function can't raise `Not_found`. Also, the name `instantiate_*` is not great. Maybe it should be `add_*`.

Reply:
well, "instantiating" a constraint is a common idiom, isn't it?

Reply:
you usually "check" a constraint

Reply:
Yes, but here I think we are "adding" a constraint to a set of constraints with these functions.

Reply:
With free variables constraints, it is the case.  Not with convertibility constraints.

Reply:
OK, I see.